server/webapp/WEB-INF/rails/webpack/models/elastic_profiles/types.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as _ from "lodash";
+import {Stream} from "mithril/stream";
+import * as stream from "mithril/stream";
+import {Errors} from "models/mixins/errors";
+import {applyMixins} from "models/mixins/mixins";
+import {ValidatableMixin} from "models/mixins/new_validatable_mixin";
+import {Configurations} from "models/shared/configuration";
+
+export class ElasticProfiles {
+  private readonly profiles: Stream<ElasticProfile[]>;
+
+  constructor(profiles: ElasticProfile[]) {
+    this.profiles = stream(profiles);
+  }
+
+  public static fromJSON(profilesJson: any): ElasticProfiles {
+    const profiles = profilesJson.map((profile: any) => {
+      return ElasticProfile.fromJSON(profile);
+    });
+    return new ElasticProfiles(profiles);
+  }
+
+  public static instance(id: string, pluginId: string) {
+    return new ElasticProfile(id, pluginId, new Configurations([]));
+  }
+
+  public all(): Stream<ElasticProfile[]> {
+    return this.profiles;
+  }
+
+  public size(): number {
+    return this.profiles() == null ? 0 : this.profiles().length;
+  }
+
+  public empty(): boolean {
+    return this.size() === 0;
+  }
+
+  public groupByPlugin() {
+    return _.groupBy(this.profiles(), (profile) => {
+      return profile.pluginId;
+    });
+  }
+}
+
+export class ProfileUsage {
+  public pipelineName: string;
+  public jobName: string;
+  public stageName: string;
+
+  constructor(pipelineName: string, stageName: string, jobName: string) {
+    this.pipelineName = pipelineName;
+    this.stageName    = stageName;
+    this.jobName      = jobName;
+  }
+
+  public static fromJSON(usageJson: any) {
+    return new ProfileUsage(usageJson.pipeline_name, usageJson.stage_name, usageJson.job_name);
+  }
+}
+
+//tslint:disable-next-line
+export interface ElasticProfile extends ValidatableMixin {
+}
+
+export class ElasticProfile implements ValidatableMixin {
+  public id: Stream<string>;
+  public pluginId: Stream<string>;
+  public properties: Configurations;
+
+  constructor(id: string, pluginId: string, properties: Configurations) {
```

Comment:
See types.ts in config repos for the convention for defining constructors, all args have been made optional (for ease of creating empty objects in tests)